File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null)return head;
        // if(n == 1)return head.next;
        
        ListNode newnode = reverse(head);
        if(n == 1)return reverse(newnode.next);
        
        int cnt = 1;
        ListNode curr = newnode;
        while(curr != null && cnt < n - 1){
            cnt++;
            curr = curr.next;
        }
        
        if(curr != null && curr.next != null && cnt == n - 1){
            curr.next = curr.next.next;
        }

        return reverse(newnode);
    }
    
    public ListNode reverse(ListNode head ){
        ListNode prev = null;
        ListNode next = null;
        ListNode curr = head;
        
        while(curr != null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
}

class Solution1 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
    if(head == null)return null;   
        
    ListNode fast = head;
    ListNode slow = head;
    for(int i = 0; i<n; i++){
        fast = fast.next;
        }    
        
     while(fast.next != null){
         fast = fast.next;
         slow = slow.next;
        }   
    
        if (fast == null)return head.next;
        
     slow.next = slow.next.next;
        
        return head;
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverse(self , head):
        curr = head
        prev = None 
        nextnode = None
        
        while curr != None:
            nextnode = curr.next
            curr.next = prev
            prev = curr
            curr = nextnode
            
        return prev
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head is None:
            return head
        
        head = self.reverse(head)
        if n == 1:
            head = head.next
            return self.reverse(head)
        
        curr = head
        cnt = 1
        
        while curr != None and cnt < n - 1:
            cnt += 1
            curr = curr.next
            
        if curr and curr.next:
            curr.next = curr.next.next
        
        return self.reverse(head)
    
        
        
        
        
        